src/Interpreters/TranslateQualifiedNamesVisitor.cpp

```diff
@@ -225,6 +227,69 @@ void TranslateQualifiedNamesMatcher::visit(ASTExpressionList & node, const ASTPt
         }
     }
 
+    auto transform = [&](const ASTPtr & transformer)
+    {
+        if (const auto * apply = transformer->as<ASTColumnsApplyTransformer>())
```

Comment:
I would like to see it with interfaces and factories...
As long as we have only three transformers, current implementation is also Ok but doesn't look good.